packages/api/src/util.ts

```diff
@@ -28,21 +28,48 @@ export const sleep = (duration) => {
 };
 
 export const semaphore = () => {
-  let resolvePromise;
-  const promise = new Promise((resolve) => {
-    resolvePromise = resolve;
-  });
-  return {
-    release: () => {
-      resolvePromise();
-    },
-    wait: (timeoutMs) => {
-      if (!timeoutMs) return promise;
-      return Promise.race([promise, sleep(timeoutMs)]);
-    },
-  };
+  const sem = new Semaphore(1);
+  sem.wait(); // semaphore from this function starts acquired
+  return sem;
 };
 
+export class Semaphore {
```

Comment:
Did you consider using some library for this instead of implementing the Semaphore?

If we end up having our own Semaphore implementation, then how about writing some unit tests?

Reply:
Not necessary anymore given I'm using a separate conn pool instead :)

But FTR, I remember I _did_ search for libraries back then and surprisingly there weren't any suitable! I guess people don't really care about concurrency in node since it's single threaded 🤷‍♂️